core/commands/mfs/mfs.go

```diff
@@ -0,0 +1,689 @@
+	path "github.com/ipfs/go-ipfs/path"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+	u "github.com/ipfs/go-ipfs/util"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+)
+
+var log = u.Logger("cmds/mfs")
+
+type mfsMountListing struct {
+	Mounts []mfs.RootListing
+}
+
+var MfsCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Manipulate mutable filesystems",
+		ShortDescription: `
+Mfs is an API for manipulating ipfs objects as if they were a unix filesystem.
+They can be seeded with an initial root hash, or by default are an empty directory.
+
+'Sessions' may be created with the create command. Once created, they show up in
+the output of the base mfs command.
+
+NOTICE:
+This API is currently experimental, likely to change, and may potentially be
+unstable. This notice will be removed when that is no longer the case. Feedback
+on how this API could be improved is very welcome.
+		`,
+	},
+	Options: []cmds.Option{
+		cmds.StringOption("s", "session", "the name of the filesystem to operate on (default=local)"),
+	},
+	Subcommands: map[string]*cmds.Command{
+		"create": MfsCreateCmd,
+		"close":  MfsCloseCmd,
+		"put":    MfsPutCmd,
+		"read":   MfsReadCmd,
+		"write":  MfsWriteCmd,
+		"mv":     MfsMvCmd,
+		"ls":     MfsLsCmd,
+		"mkdir":  MfsMkdirCmd,
+		"rm":     MfsRmCmd,
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		node, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		if node.Mfs == nil {
+			res.SetError(cmds.ErrNotOnline, cmds.ErrNormal)
+			return
+		}
+
+		session, found, err := req.Option("session").String()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+		if found {
+			// if a session is specified, print out just the root hash for that fs
+			root, err := node.Mfs.GetRoot(session)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			nd, err := root.GetValue().GetNode()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			k, err := nd.Key()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			res.SetOutput(&mfsMountListing{[]mfs.RootListing{mfs.RootListing{Hash: k}}})
+			return
+		}
+
+		// Default: list all roots
+		roots := node.Mfs.ListRoots()
+
+		res.SetOutput(&mfsMountListing{roots})
+	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			out := res.Output().(*mfsMountListing)
+			if len(out.Mounts) == 1 && out.Mounts[0].Name == "" {
+				// a session was specified, print out just the hash
+				return strings.NewReader(out.Mounts[0].Hash.B58String()), nil
+			}
+
+			buf := new(bytes.Buffer)
+			for _, o := range out.Mounts {
+				fmt.Fprintf(buf, "%s - %s\n", o.Name, o.Hash)
+			}
+			return buf, nil
+		},
+	},
+	Type: mfsMountListing{},
+}
+
+var MfsCreateCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Create a new mutable filesystem",
```

Comment:
Hmm not sure this should be called "a filesystem". It's confusing to think about "using ipfs  -- a distributed file system -- to then have to create other filesystems-- what?" (i know it's technically a filesystem, but the overloading is hard on the user)

Maybe this should be called "a mutable root" or "a mutable space", or just "a mutable dag"?

Another option is to, instead of calling it a thing and letting users create roots, maybe we should:

- have a proper root object that links to all these "roots"
- users mutate _that_ root always, and they get only one
- in git-land, consider a repo. a repo is really just a "tree" with links, whose names are "branches" and values are "commit" hashes.
- likewise here we could have the root be the user's "local" data, every root of an "ipfs add" gets added to a node under that called "/local/added" (like a log) (( maybe this is exactly what already happens )) 
- and "mfs create", would be the equivalent of "ipfs local link <name> <root>"  ?

---

- we could also name these  "branches" (bad if we want to do proper commit branches later) or 
- "views" (like a "view on the data")